pgserver/connection_handler.go

```diff
@@ -1409,14 +1406,31 @@ func (h *ConnectionHandler) handleCopyToStdout(query ConvertedQuery, copyTo *tre
 		for {
 			n, err := pipe.Read(buf)
 			if n > 0 {
-				copyData := &pgproto3.CopyData{
-					Data: buf[:n],
+				sendCopyData := func(copyData *pgproto3.CopyData) {
+					ctx.GetLogger().Debugf("sending CopyData (%d bytes) to the client", n)
+					if err := h.send(copyData); err != nil {
+						sendErr.Store(err)
+						cancel()
+						return
+					}
 				}
-				ctx.GetLogger().Debugf("sending CopyData (%d bytes) to the client", n)
-				if err := h.send(copyData); err != nil {
-					sendErr.Store(err)
-					cancel()
-					return
+				switch format {
+				case tree.CopyFormatText:
```

Comment:
Interesting! It would be better to use `bufio.Reader` and its [ReadSlice](https://pkg.go.dev/bufio#Reader.ReadSlice) method here for best performance.